crates/evm/src/execute.rs

```diff
@@ -14,33 +19,123 @@ pub use reth_storage_errors::provider::ProviderError;
 ///
 /// The trait cannot commit the state changes to the database directly, see [`EvmCommit`].
 pub trait EvmTransact {
-    /// The environment for the evm.
-    type Env;
-    /// The output produced by the evm.
-    type Output;
-    /// The error produced by the evm.
-    type Error;
+    /// The database used by the evm.
+    type DB: Database;
 
     /// Transacts with the current env to produce an output without
     /// committing to the underlying database.
-    fn transact(&mut self) -> Result<Self::Output, Self::Error>;
+    fn transact(&mut self) -> EVMResult<<<Self as EvmTransact>::DB as Database>::Error>;
 
     /// Returns a mutable reference to the current env.
-    fn env_mut(&mut self) -> &mut Self::Env;
+    fn env_mut(&mut self) -> &mut Env;
 
     /// Returns a reference to the current env.
-    fn env(&self) -> &Self::Env;
+    fn env(&self) -> &Env;
+
+    /// Returns the current env with handler cfg.
+    fn env_with_handler_cfg(&self) -> EnvWithHandlerCfg;
+
+    /// Returns a reference to the current database.
+    fn db(&self) -> &Self::DB;
+
+    /// Returns a mutable reference to the current database.
+    fn db_mut(&mut self) -> &mut Self::DB;
+
+    /// Returns the reference of transaction.
+    fn tx(&self) -> &TxEnv;
+
+    /// Returns the mutable reference of transaction.
+    fn tx_mut(&mut self) -> &mut TxEnv;
+
+    /// Returns the reference of block.
+    fn block(&self) -> &BlockEnv;
+}
+
+/// Trait that represents the output of a transaction and provides access to the state.
+pub trait EvmTransactOutput {
+    /// Returns the state produced by the transaction.
+    fn state(&self) -> EvmState;
 }
 
 /// Trait that can transact an [`EvmTransact::Env`] and
 /// commit state changes to the database.
 pub trait EvmCommit: EvmTransact {
+    /// The output produced by the evm.
+    type CommitOutput;
+    /// The error produced by the evm.
+    type CommitError;
+
     /// Commit state changes to the database.
-    fn commit(&mut self, output: Self::Output);
+    fn commit(&mut self, output: HashMap<Address, Account>);
```

Comment:
this should be `Self::CommitOutput` no?

Reply:
I do not think so. I just changed the naming because it was indeed a bit confusing but the output of commit is not the same as the output of transact and commit.

Indeed in commit, I think we can say that in a fairly generic way (true for any implementation of evm), the output must be a hashmap of address and account after the commit.

On the other hand, the output of transact and commit is an `ExecutionResult` (success, revert, halt).

Reply:
but you could argue that the commit can take the evm transact output `ResultAndState` and ingest this in order to commit the state changes to the db. This makes the function more generic and we already have the `ResultAndState` struct.

Reply:
You are right on this. I just used the `ResultAndState` as the commit argument. But I didn't add a generic type for TransactOutput. It added one more generic and a little complexity. Let's keep this for a follow up in order to restrict the scope as much as possible here knowing that it is not hyper blocking.